src/ripple/app/consensus/RCLCensorshipDetector.h

```diff
@@ -0,0 +1,128 @@
+
+#include <ripple/shamap/SHAMap.h>
+#include <algorithm>
+#include <map>
+
+namespace ripple {
+
+template <class TxID, class Sequence>
+class RCLCensorshipDetector
+{
+private:
+    std::map<TxID, Sequence> tracker;
+
+public:
+    RCLCensorshipDetector() = default;
+
+    /** Add transactions being proposed for the current consensus round.
+
+        @param seq      The sequence number of the ledger being built.
+        @param proposed The set of transactions that we are initially proposing
+                        for this round.
+    */
+    void propose(
+        Sequence seq,
+        std::vector<TxID> proposed)
+    {
+        std::sort (proposed.begin(), proposed.end());
+
+        // We want to remove any entries that we proposed in a previous round
+        // that did not make it in yet if we are no longer proposing them.
+        prune ([&proposed](TxID const& id, Sequence seq)
+            {
+                return !std::binary_search(proposed.begin(), proposed.end(), id);
+            });
+
+        // Track the entries that we are proposing in this round.
+        for (auto const& p : proposed)
+            tracker.emplace(p, seq); // FIXME C++17: use try_emplace
+    }
+
+    /** Determine which transactions made it and perform censorship detection.
+
+        This function is called when the server is proposing and a consensus
+        round it participated in completed.
+
+        @param accepted The set of transactions that the network agreed
+                        should be included in the ledger being built.
+        @param pred     A predicate invoked for every transaction we've proposed
+                        but which hasn't yet made it. The predicate must be
+                        callable as:
+                            bool pred(TxID const&, Sequence)
+                        It must return true for entries that should be removed.
+    */
+    template <class Predicate>
+    void check(
+        std::vector<TxID> accepted,
+        Predicate pred)
+    {
+        std::sort (accepted.begin(), accepted.end());
+
+        // We want to remove all tracking entries for transactions that were
+        // accepted as well as those which match the predicate.
+        prune ([&pred, &accepted](TxID const& id, Sequence seq)
+            {
+                if (std::binary_search(accepted.begin(), accepted.end(), id))
+                    return true;
+
+                return pred(id, seq);
+            });
+    }
+
+    /** Removes all elements satisfying specific criteria from the tracker
+
+        @param pred     A predicate which returns true for tracking entries
+                        that should be removed. The predicate must be callable
+                        as:
+                            bool pred(TxID const&, Sequence)
+                        It must return true for entries that should be removed.
+    */
+    template <class Predicate>
+    void prune(Predicate pred)
```

Comment:
Make this private?

Reply:
This may be useful in different contexts (e.g. a command to display the contents of the detector) so I'd rather leave it.